File: packages/contracts/src/primitives.ts
import { z } from "zod";

/**
 * Reusable value objects and envelope types used across entities and the API.
 * These are deliberately small and storage-agnostic.
 */

/** ISO-8601 UTC timestamp. We keep timestamps as strings end-to-end so the
 *  same value crosses the JSON boundary unchanged (no Date drift). */
export const Timestamp = z.string().datetime({ offset: true });
export type Timestamp = z.infer<typeof Timestamp>;

/** A monetary amount. Stored as a number of USD for now; wrapped so a future
 *  multi-currency/minor-unit migration touches one type, not every call site. */
export const Money = z.object({
  amountUsd: z.number().nonnegative(),
});
export type Money = z.infer<typeof Money>;

/** Token accounting for a run or agent. */
export const Usage = z.object({
  tokensIn: z.number().int().nonnegative(),
  tokensOut: z.number().int().nonnegative(),
});
export type Usage = z.infer<typeof Usage>;

export const Viewport = z.object({
  width: z.number().int().positive(),
  height: z.number().int().positive(),
});
export type Viewport = z.infer<typeof Viewport>;

/** done/total step counter shown on run progress bars. */
export const StepProgress = z.object({
  done: z.number().int().nonnegative(),
  total: z.number().int().nonnegative(),
});
export type StepProgress = z.infer<typeof StepProgress>;

/** Per-agent live metrics surfaced in the graph and inspector. */
export const AgentMetrics = z.object({
  tokens: z.number().int().nonnegative(),
  costUsd: z.number().nonnegative(),
  runtimeMs: z.number().int().nonnegative().nullable(),
  toolCalls: z.number().int().nonnegative(),
  /** 0..1 */
  progress: z.number().min(0).max(1),
});
export type AgentMetrics = z.infer<typeof AgentMetrics>;

/** Person reference used for owners / triggered-by / actors. */
export const Principal = z.object({
  name: z.string(),
  initials: z.string().min(1).max(3),
});
export type Principal = z.infer<typeof Principal>;

/* ------------------------------------------------------------------ *
   Pagination — cursor-based, the only model that survives growth.
 * ------------------------------------------------------------------ */

export const PageQuery = z.object({
  /** Opaque cursor returned by a previous page. */
  cursor: z.string().optional(),
  limit: z.coerce.number().int().min(1).max(100).default(25),
});
export type PageQuery = z.infer<typeof PageQuery>;

export interface Page<T> {
  items: T[];
  /** Cursor to fetch the next page, or null at the end. */
  nextCursor: string | null;
  /** Total count when cheaply available (in-memory store); else undefined. */
  total?: number;
}

/** Build the response schema for a paginated list of `item`. */
export function pageOf<T extends z.ZodTypeAny>(item: T) {
  return z.object({
    items: z.array(item),
    nextCursor: z.string().nullable(),
    total: z.number().int().nonnegative().optional(),
  });
}

File: apps/api/internal/domain/primitives.go
package domain

// Reusable value objects and envelope types used across entities and the API.
// Storage-agnostic and deliberately small, mirroring
// packages/contracts/src/primitives.ts.

// Timestamp is an ISO-8601 UTC timestamp kept as a string end-to-end so the
// same value crosses the JSON boundary unchanged (no Date drift).
type Timestamp string

// Usage is token accounting for a run or agent.
type Usage struct {
	TokensIn  int `json:"tokensIn"`
	TokensOut int `json:"tokensOut"`
}

type Viewport struct {
	Width  int `json:"width"`
	Height int `json:"height"`
}

// StepProgress is the done/total step counter shown on run progress bars.
type StepProgress struct {
	Done  int `json:"done"`
	Total int `json:"total"`
}

// AgentMetrics are per-agent live metrics surfaced in the graph and inspector.
type AgentMetrics struct {
	Tokens    int     `json:"tokens"`
	CostUsd   float64 `json:"costUsd"`
	RuntimeMs *int    `json:"runtimeMs"` // nullable
	ToolCalls int     `json:"toolCalls"`
	Progress  float64 `json:"progress"` // 0..1
}

// Principal is a person reference used for owners / triggered-by / actors.
type Principal struct {
	Name     string `json:"name"`
	Initials string `json:"initials"`
}

// Page is a cursor-paginated list response — the only model that survives growth.
type Page[T any] struct {
	Items      []T     `json:"items"`
	NextCursor *string `json:"nextCursor"`
	Total      *int    `json:"total,omitempty"`
}

// PageQuery is the shared cursor-based pagination query.
type PageQuery struct {
	Cursor string // opaque cursor returned by a previous page
	Limit  int    // 1..100, default 25
}
